icicle/include/icicle/program/symbol.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <memory>
+#include <vector>
+
+namespace icicle {
+
+  /**
+   * @brief Enum to represent all the operation supported by the Symbol class.
+   *
+   */
+  enum OpCode {
+    OP_ADD = 0,
+    OP_MULT,
+    OP_SUB,
+    OP_INV,
+
+    NOF_OPERATIONS,
+
+    OP_INPUT,
+    OP_CONST,
+  };
+
+  /**
+   * @brief a node at the data-flow grapsh (DFG) that represent a single operation
+   *
+   * This class contains a data about an operation done by the Symbol class.
+   * It saves oll the data to describe the operation done and point to its operands.
+   * At any given time a Sybol contains an Operation instance that represent all the
+   * calculation history done so far to reach the current value.
+   */
+
+  template <typename S>
+  class Operation
+  {
+  public:
+    OpCode m_opcode;
+    std::shared_ptr<Operation<S>> m_operand1; // 1st operand if exist
+    std::shared_ptr<Operation<S>> m_operand2; // 2nd operand if exist
+
+    // optinal parameters:
+    std::unique_ptr<S> m_constant; // for OP_CONST: const value
+
+    // implementation:
+    int m_mem_addr; // location at the input vectors
```

Comment:
should this be here? this seems like it is coupling this class to something else. Also int may not be enough for large data. probably u64 is better

Reply:
The name might be misleading. 
This is the idx in the variable vector that this operation is going to store the result.
This is relevant to Program and not the executor.